src/network/netdev/macsec.c

```diff
@@ -0,0 +1,928 @@
+
+static int macsec_transmit_transmit_association_new_static(MACsec *s, const char *filename, unsigned section_line, TransmitAssociation **ret) {
+        _cleanup_(network_config_section_freep) NetworkConfigSection *n = NULL;
+        _cleanup_(macsec_transmit_association_freep) TransmitAssociation *a = NULL;
+        int r;
+
+        assert(s);
+        assert(ret);
+        assert(filename);
+        assert(section_line > 0);
+
+        r = network_config_section_new(filename, section_line, &n);
+        if (r < 0)
+                return r;
+
+        a = ordered_hashmap_get(s->transmit_association_by_section, n);
+        if (a) {
+                *ret = TAKE_PTR(a);
+                return 0;
+        }
+
+        a = new(TransmitAssociation, 1);
+        if (!a)
+                return -ENOMEM;
+
+        *a = (TransmitAssociation) {
+                .mac_sec = s,
+                .section = TAKE_PTR(n),
+        };
+
+        r = ordered_hashmap_ensure_allocated(&s->transmit_association_by_section, &network_config_hash_ops);
+        if (r < 0)
+                return r;
+
+        r = ordered_hashmap_put(s->transmit_association_by_section, a->section, a);
+        if (r < 0)
+                return r;
+
+        *ret = TAKE_PTR(a);
+
+        return 0;
+}
+
+static int netdev_macsec_fill_message(NetDev *netdev, int command, sd_netlink_message **ret) {
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL;
+        MACsec *t;
+        int r;
+
+        assert(netdev);
+        assert(netdev->ifindex > 0);
+
+        t = MACSEC(netdev);
+
+        assert(t);
+
+        r = sd_genl_message_new(netdev->manager->genl, SD_GENL_MACSEC, command, &m);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Failed to create generic netlink message: %m");
+
+        r = sd_netlink_message_append_u32(m, MACSEC_ATTR_IFINDEX, netdev->ifindex);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append MACSEC_ATTR_IFINDEX attribute: %m");
+
+        *ret = TAKE_PTR(m);
+
+        return 0;
+}
+
+static int netdev_macsec_fill_message_sci(NetDev *netdev, ReceiveChannel *c, sd_netlink_message *m) {
+        union {
+                uint64_t as_uint64;
+
+                struct {
+                        struct ether_addr addr;
+                        uint16_t port;
+                } as_address_and_port;
+        } sci;
+        MACsec *t;
+        int r;
+
+        assert(netdev);
+        assert(m);
+
+        t = MACSEC(netdev);
+
+        assert(t);
+
+        assert(c);
+        assert(c->port > 0);
+
+        r = sd_netlink_message_open_container(m, MACSEC_ATTR_RXSC_CONFIG);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append MACSEC_ATTR_RXSC_CONFIG attribute: %m");
+
+        memcpy(&sci.as_uint64, &c->mac, ETH_ALEN);
+        memcpy(((char *) &sci.as_uint64) + ETH_ALEN, &c->port, sizeof(c->port));
```

Comment:
Then now, this can be just a normal assignment: `sci.as_address_and_port.port = c->port;`

Reply:
I have not tested this. Looking at the kernel code they do this in same manner. It also validates that.

Reply:
that works. replaced ;